lib/util/removeUnnecessaryParens.js

```diff
@@ -1,174 +1,53 @@
-  // Parens that wrap everything are redundant.
-  // NOTE: removeUnnecessaryParensSearch recursively removes parens that aren't
-  // needed, while this step only applies to the very top level expression.
-  // e.g. (2 + 3) * 4 can't become 2 + 3 * 4, but if (2 + 3) as a top level
-  // expression can become 2 + 3
-  if (rootNode) {
-    while (Node.Type.isParenthesis(node)) {
-      node = node.content;
-    }
-  }
-  return removeUnnecessaryParensSearch(node);
-}
-
-// Recursively moves parenthesis around nodes that can't be resolved further if
-// it doesn't change the value of the expression. Returns a node.
-// NOTE: after this function is called, every parenthesis node in the
-// tree should always have an operator node or unary minus as its child.
-function removeUnnecessaryParensSearch(node) {
-  if (Node.Type.isOperator(node)) {
-    return removeUnnecessaryParensInOperatorNode(node);
-  }
-  else if (Node.Type.isFunction(node)) {
-    return removeUnnecessaryParensInFunctionNode(node);
-  }
-  else if (Node.Type.isParenthesis(node)) {
-    return removeUnnecessaryParensInParenthesisNode(node);
-  }
-  else if (Node.Type.isConstant(node, true) || Node.Type.isSymbol(node)) {
-    return node;
-  }
-  else if (Node.Type.isUnaryMinus(node)) {
-    const content = node.args[0];
-    node.args[0] = removeUnnecessaryParensSearch(content);
-    return node;
-  }
-  else {
-    throw Error('Unsupported node type: ' + node.type);
-  }
-}
-
-// Removes unncessary parens for each operator in an operator node, and removes
-// unncessary parens around operators that can't be simplified further.
-// Returns a node.
-function removeUnnecessaryParensInOperatorNode(node) {
-  // Special case: if the node is an exponent node and the base
-  // is an operator, we should keep the parentheses for the base.
-  // e.g. (2x)^2 -> (2x)^2 instead of 2x^2
-  if (node.op === '^' && Node.Type.isParenthesis(node.args[0])) {
-    const base = node.args[0];
-    if (Node.Type.isOperator(base.content)) {
-      base.content = removeUnnecessaryParensSearch(base.content);
-      node.args[1] = removeUnnecessaryParensSearch(node.args[1]);
-
-      return node;
-    }
-  }
-
-  node.args.forEach((child, i) => {
-    node.args[i] = removeUnnecessaryParensSearch(child);
-  });
-
-  // Sometimes, parens are around expressions that have been simplified
-  // all they can be. If that expression is part of an addition or subtraction
-  // operation, we can remove the parenthesis.
-  // e.g. (x+4) + 12 -> x+4 + 12
-  if (node.op === '+') {
-    node.args.forEach((child, i) => {
-      if (Node.Type.isParenthesis(child) &&
-          !canCollectOrCombine(child.content)) {
-        // remove the parens by replacing the child node (in its args list)
-        // with its content
-        node.args[i] = child.content;
+// Note: this function does more than remove unnecessary parens, it also
+// flattens nested expressions, e.g. (1 + x) + (2 + y), sets wasMinus = true on
```

Comment:
:(  I should've wrote `(1 + x) + (2 + y) --> 1 + x + 2 + y`... I'll update this.